### PR TITLE
Add tracing for solr datasource

### PR DESCRIPTION
**Pull request by Umang01-hash**
## Pull Request Template

**Description:**

-   Add custom spans to support tracing in solr datasource.

Screenshot 2024-10-15 at 1 08 36 PM

**Checklist:**

-   [x] I have formatted my code using  `goimport`  and  `golangci-lint`.
-   [x] All new code is covered by unit tests.
-   [x] This PR does not decrease the overall code coverage.
-   [x] I have reviewed the code comments and documentation for clarity.

**Thank you for your contribution!**

### CODE REVIEW

File: pkg/gofr/datasource/solr/solr.go [hunk 1]
```diff
@@ -168,7 +196,52 @@ type Response struct {
...
+
+	if c.tracer != nil {
+		ctx, span = c.tracer.Start(ctx, fmt.Sprintf("Solr %s", method),
+			trace.WithAttributes(
+				attribute.String("solr.url", url),
+			),
+		)
+	}
+
+	req, err := c.createRequest(ctx, method, url, params, body)
```
Comment (vipul-rawat) [inline pkg/gofr/datasource/solr/solr.go:210]: we can use `httptrace` package to trace the underlying requests and have the same context inherit the traceId from the parent call.

File: pkg/gofr/datasource/solr/solr.go [hunk 2]
```diff
@@ -193,36 +266,25 @@ func call(ctx context.Context, method, url string, params map[string]any, body i
 
 	req.URL.RawQuery = q.Encode()
 
-	client := &http.Client{}
+	req = req.WithContext(ctx)
```
Comment (vipul-rawat) [inline pkg/gofr/datasource/solr/solr.go:269]: we can directly create the req with context on line 245

File: pkg/gofr/datasource/solr/solr.go [hunk 3]
```diff
@@ -168,7 +196,52 @@ type Response struct {
...
+			),
+		)
+	}
+
+	req, err := c.createRequest(ctx, method, url, params, body)
+	if err != nil {
+		return nil, err, nil
+	}
+
+	client := &http.Client{}
```
Comment (vipul-rawat) [inline pkg/gofr/datasource/solr/solr.go:215]: we can avoid creating a new http client for every call, instead have it in the `Client`  struct itself

File: pkg/gofr/datasource/solr/solr.go (hunk 1 above)
Comment (Umang01-hash) [reply pkg/gofr/datasource/solr/solr.go:216]: sure done

File: pkg/gofr/datasource/solr/solr.go (hunk 3 above)
Comment (Umang01-hash) [reply pkg/gofr/datasource/solr/solr.go]: sure did the same

File: pkg/gofr/datasource/solr/solr.go (hunk 2 above)
Comment (Umang01-hash) [reply pkg/gofr/datasource/solr/solr.go]: sure will do

**Review by vipul-rawat — APPROVED** [review]

**Pull request merged**